src/cockpit/channels/dbus.py

```diff
@@ -0,0 +1,204 @@
+        except KeyError:
+            pass
+
+        if bus and destination and object_path:
+            try:
+                await self.introspect_path(bus, destination, object_path)
+            except BusError:
+                pass
+
+        return self.cache.get(interface_name)
+
+    async def get_signature(self, interface_name, method, bus=None, destination=None, object_path=None):
+        interface = await self.get_interface(interface_name, bus, destination, object_path)
+        if interface is None:
+            raise KeyError(f'Interface {interface_name} is not found')
+
+        return ''.join(interface['methods'][method]['in'])
+
+
+class DBusChannel(Channel):
+    payload = 'dbus-json3'
+
+    tasks = None
+    matches = None
+    name = None
+    bus = None
+
+    def do_open(self, options):
+        self.cache = InterfaceCache()
+        self.name = options['name']
+        self.matches = []
+        self.tasks = set()
+
+        if options.get('bus') == 'session':
+            logger.debug('get session bus for %s', self.name)
+            self.bus = Bus.default_user()
+        else:
+            logger.debug('get system bus for %s', self.name)
+            self.bus = Bus.default_system()
+
+        try:
+            self.bus.attach_event(None, 0)
+        except OSError as err:
+            if err.errno != errno.EBUSY:
+                raise
+
+        self.ready()
+
+    def match_hit(self, message):
+        logger.debug('got match')
+        self.send_message(signal=[
+            message.get_path(),
+            message.get_interface(),
+            message.get_member(),
+            list(message.get_body())
+        ])
+
+    async def do_call(self, call, message):
+        path, iface, method, args = call
+        timeout = message.get('timeout')
+        cookie = message.get('id')
+        flags = message.get('flags')
+
+        # We have to figure out the signature of the call.  Either we got told it:
+        signature = message.get('type')
+
+        # ... or all of our arguments are string (and we can guess):
+        if signature is None and all(isinstance(arg, str) for arg in args):
+            signature = 's' * len(args)
+
+        # ... or we need to introspect
+        try:
+            logger.debug('Doing introspection request for %s %s', iface, method)
+            signature = await self.cache.get_signature(iface, method, self.bus, self.name, path)
+        except BusError as error:
+            self.send_message(error=[error.code, [f'Introspection: {error.description}']], id=cookie)
+            return
+        except Exception as exc:
+            self.send_message(error=['python.error', [f'Introspection: {str(exc)}']], id=cookie)
+            return
+
+        try:
+            reply = await self.bus.call_method_async(self.name, path, iface, method, signature, *args, timeout=timeout)
+            self.send_message(reply=[reply], id=cookie, flags="<" if flags is not None else None)
```

Comment:
We should get the endianess right here, the NetworkManager API actually requires us to know.  (Or used to, not sure if it still does.)

Reply:
I'll add a TODO for this.